include/swift/Basic/LangOptions.h

```diff
@@ -585,6 +585,12 @@ namespace swift {
     /// than this many seconds.
     unsigned ExpressionTimeoutThreshold = 600;
 
+    /// If the shared pointer is not a \c nullptr and the pointee is \c true,
+    /// typechecking should be aborted at the next possible opportunity.
+    /// This is used by SourceKit to cancel requests for which the result is no
+    /// longer of interest.
+    std::shared_ptr<bool> CancellationFlag = nullptr;
```

Comment:
I thought about race conditions here a bit and *think* it should be fine with a plain pointer. We should only have one location setting the cancellation flag to `true` and that’s independent of what it’s previous content was, so there shouldn’t be any write races. And if setting the flag to `true` doesn’t immediately propagate to the thread that is building the AST, it’s also fine since nobody expects the cancellation to be immediate – it’s totally fine for the type checker to e.g. continue type checking for a few more cycles. But I might be missing something.

But we could of course make it a `std::atomic`. Not quite sure if that will add some performance overhead though…

Reply:
We cannot ignore the read-write race:
* The compiler is free optimize this code based on the assumption no other thread modifies the value, since it's not valid to do so.
* TSan will raise an error on this race.

> But we could of course make it a std::atomic. Not quite sure if that will add some performance overhead though…

Using relaxed memory order atomic operations shouldn't cost much. On x86_64 it should be basically free.

Reply:
OK, I changed it to a `std::atomic`. 

Out of curiosity though: How would the race condition manifest at runtime except for the TSan failure?